File: app/src/main/java/com/irateam/vkplayer/api/service/SettingsService.java
package com.irateam.vkplayer.api.service;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Environment;
import android.preference.PreferenceManager;

import com.irateam.vkplayer.player.Player;
import com.irateam.vkplayer.services.DownloadService;

import java.io.File;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

public final class SettingsService {

    private static final int SYNC_ALARM_ID = 1;
    private static final int DEFAULT_SYNC_COUNT = 10;

    public static final String REPEAT_STATE = "repeat_state";
    public static final String RANDOM_STATE = "random_state";
    public static final String SYNC_ENABLED = "sync_enabled";
    public static final String SYNC_TIME = "sync_time";
    public static final String SYNC_COUNT = "sync_count";
    public static final String SYNC_WIFI = "sync_wifi";

    private final Context context;
    private final SharedPreferences preferences;

    private static SettingsService instance;

    public static synchronized SettingsService getInstance(Context context) {
        if (instance == null) {
            instance = new SettingsService(context);
        }
        return instance;
    }

    private SettingsService(Context context) {
        this.context = context;
        this.preferences = PreferenceManager.getDefaultSharedPreferences(context);
    }


    public void setPlayerRepeat(Player.RepeatState state) {
        preferences.edit()
                .putString(REPEAT_STATE, state.name())
                .apply();
    }

    public Player.RepeatState getPlayerRepeat() {
        return Player.RepeatState.valueOf(preferences.getString(REPEAT_STATE, Player.RepeatState.NO_REPEAT.name()));
    }

    public void setRandomState(boolean state) {
        preferences.edit()
                .putBoolean(RANDOM_STATE, state)
                .apply();
    }

    public boolean getRandomState() {
        return preferences.getBoolean(RANDOM_STATE, false);
    }

    public void setSyncEnabled(boolean enabled) {
        preferences.edit()
                .putBoolean(SYNC_ENABLED, enabled)
                .apply();
    }

    public boolean isSyncEnabled() {
        return preferences.getBoolean(SYNC_ENABLED, false);
    }

    public void setSyncWifi(boolean isWifi) {
        preferences.edit()
                .putBoolean(SYNC_WIFI, isWifi)
                .apply();
    }

    public boolean isWifiSync() {
        return preferences.getBoolean(SYNC_WIFI, false);
    }

    public void setSyncTime(int hour, int minutes) {
        preferences.edit()
                .putString(SYNC_TIME, String.format("%02d", hour) + ":" + String.format("%02d", minutes))
                .apply();
    }

    public Calendar getSyncTime() {
        Calendar calendar = Calendar.getInstance();
        Date date = null;
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("HH:mm");
        try {
            date = simpleDateFormat.parse(preferences.getString(SYNC_TIME, "18:30"));
        } catch (ParseException e) {
            e.printStackTrace();
        }

        calendar.set(Calendar.HOUR_OF_DAY, date.getHours());
        calendar.set(Calendar.MINUTE, date.getMinutes());
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        if (calendar.getTimeInMillis() < System.currentTimeMillis()) {
            calendar.set(Calendar.DAY_OF_YEAR, calendar.get(Calendar.DAY_OF_YEAR) + 1);
        }
        return calendar;
    }

    public void setSyncCount(int count) {
        preferences.edit()
                .putString(SYNC_COUNT, String.valueOf(count))
                .apply();
    }

    public int getSyncCount() {
        if (preferences.getString(SYNC_COUNT, "").isEmpty())
            return DEFAULT_SYNC_COUNT;
        else
            return Integer.valueOf(preferences.getString(SYNC_COUNT, String.valueOf(DEFAULT_SYNC_COUNT)));
    }

    public File getAudioCacheDir() {
        String state = Environment.getExternalStorageState();
        File cacheDir;
        if (Environment.MEDIA_MOUNTED.equals(state)) {
            cacheDir = context.getExternalFilesDir(null);
        } else {
            cacheDir = context.getFilesDir();
        }
        return cacheDir;
    }

    public static void setSyncAlarm(Context context) {
        Intent intent = DownloadService.startSyncIntent(context, false);
        PendingIntent pendingIntent = PendingIntent.getService(context,
                SYNC_ALARM_ID,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT);

        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        alarmManager.setRepeating(AlarmManager.RTC_WAKEUP,
                SettingsService.getInstance(context).getSyncTime().getTimeInMillis(),
                AlarmManager.INTERVAL_DAY,
                pendingIntent);
    }

    public static void cancelSyncAlarm(Context context) {
        Intent intent = DownloadService.startSyncIntent(context, false);
        PendingIntent pendingIntent = PendingIntent.getService(context,
                SYNC_ALARM_ID,
                intent,
                0);

        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        alarmManager.cancel(pendingIntent);
    }
}

File: app/src/main/java/com/irateam/vkplayer/api/service/SettingsService.kt
package com.irateam.vkplayer.api.service

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.SharedPreferences
import android.os.Environment
import android.preference.PreferenceManager
import com.irateam.vkplayer.player.Player.RepeatState
import com.irateam.vkplayer.service.DownloadService
import com.irateam.vkplayer.util.SharedPreferencesProvider
import com.irateam.vkplayer.util.extension.SharedPreferencesDelegates.boolean
import com.irateam.vkplayer.util.extension.SharedPreferencesDelegates.custom
import com.irateam.vkplayer.util.extension.SharedPreferencesDelegates.int
import com.irateam.vkplayer.util.extension.SharedPreferencesDelegates.string
import com.irateam.vkplayer.util.extension.SharedPreferencesDelegates.time
import java.io.File
import java.util.*

class SettingsService : SharedPreferencesProvider {

    private val context: Context
    override val sharedPreferences: SharedPreferences

    constructor(context: Context) {
        this.context = context
        this.sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
    }

    var language: String by string(getCurrentLanguageCode())

    //Playback
    var repeatState: RepeatState by custom(RepeatState.NO_REPEAT, { RepeatState.valueOf(it) })
    var randomState: Boolean by boolean(false)

    //Sync
    var syncEnabled: Boolean by boolean(false)
    var syncWifiOnly: Boolean by boolean(false)
    var syncCount: Int by int(DEFAULT_SYNC_COUNT)
    var syncTime: Calendar by time(DEFAULT_SYNC_TIME)

    fun getAudioCacheDir(): File {
        val state = Environment.getExternalStorageState()
        return if (Environment.MEDIA_MOUNTED == state) {
            context.getExternalFilesDir(null)
        } else {
            context.filesDir
        }
    }

    fun setSyncAlarm() {
        val intent = DownloadService.startSyncIntent(context, false)
        val pendingIntent = PendingIntent.getService(context,
                SYNC_ALARM_ID,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT)

        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        alarmManager.setRepeating(AlarmManager.RTC_WAKEUP,
                syncTime.timeInMillis,
                AlarmManager.INTERVAL_DAY,
                pendingIntent)
    }

    fun cancelSyncAlarm() {
        val intent = DownloadService.startSyncIntent(context, false)
        val pendingIntent = PendingIntent.getService(context,
                SYNC_ALARM_ID,
                intent,
                0)
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        alarmManager.cancel(pendingIntent)
    }

    private fun getCurrentLanguageCode() : String {
        return Locale.getDefault().language
    }

    companion object {

        private val SYNC_ALARM_ID = 1
        private val DEFAULT_SYNC_COUNT = 10
        private val DEFAULT_SYNC_TIME: Long = 60 * 60 * 1000

        @JvmField val SYNC_TIME = "sync_time"
    }
}